src/main/java/org/ibp/api/rest/labelprinting/ObservationDatasetLabelPrinting.java

```diff
@@ -0,0 +1,562 @@
+			.put(studyTransactionsDto.getObservationUnits().get(0).getObsUnitId(), studyTransactionsDto));
+
+		final Map<Integer, Field> termIdFieldMap = Maps.uniqueIndex(labelsGeneratorInput.getAllAvailablefields(), Field::getId);
+
+		final Set<Integer> allRequiredKeys = new HashSet<>();
+		if (labelsGeneratorInput.isBarcodeRequired()) {
+			if (labelsGeneratorInput.isAutomaticBarcode()) {
+				allRequiredKeys.add(ObsDatasetUnitIdFieldKey);
+			} else {
+				allRequiredKeys.addAll(labelsGeneratorInput.getBarcodeFields());
+			}
+		}
+		labelsGeneratorInput.getFields().forEach(f -> allRequiredKeys.addAll(f));
+
+		final Map<String, String> gidPedigreeMap = new HashMap<>();
+
+		final List<ObservationUnitRow> observationUnitRows =
+			this.middlewareDatasetService.getAllObservationUnitRows(labelsGeneratorInput.getStudyId(), labelsGeneratorInput.getDatasetId());
+
+		Collections.sort(
+			observationUnitRows,
+			Comparator.comparing((ObservationUnitRow o) -> Integer.valueOf(o.getVariables().get(TRIAL_INSTANCE).getValue()))
+				.thenComparing(o -> Integer.valueOf(o.getVariables().get(PLOT_NO).getValue()))
+				.thenComparing(o -> Integer.valueOf(o.getVariables().get(ENTRY_NO).getValue())));
+
+		// Data to be exported
+		final List<Map<Integer, String>> results = new LinkedList<>();
+
+		for (final ObservationUnitRow observationUnitRow : observationUnitRows) {
+			final Map<Integer, String> row = new HashMap<>();
+			for (final Integer requiredField : allRequiredKeys) {
+				final Field field = termIdFieldMap.get(requiredField);
+				if (!STATIC_FIELD_IDS.contains(field.getId())) {
+					// Special cases: LOCATION_NAME, PLOT OBS_UNIT_ID, CROP_SEASON_CODE
+					final Integer termId = requiredField;
+					if (TermId.getById(termId).equals(TermId.LOCATION_ID)) {
+						row.put(requiredField, observationUnitRow.getVariables().get(LOCATION_ID).getValue());
+						continue;
+					}
+					if (TermId.getById(termId).equals(TermId.OBS_UNIT_ID)) {
+						row.put(requiredField, observationUnitRow.getVariables().get(OBS_UNIT_ID).getValue());
+						continue;
+					}
+					if (TermId.getById(termId).equals(TermId.SEASON_VAR)) {
+						final ObservationUnitData observationUnitData =
+							observationUnitRow.getEnvironmentVariables().get("Crop_season_Code");
+						row.put(requiredField, this.getSeason(observationUnitData != null ? observationUnitData.getValue() : null));
+						continue;
+					}
+					if (observationUnitRow.getVariables().containsKey(field.getName())) {
+						row.put(requiredField, observationUnitRow.getVariables().get(field.getName()).getValue());
+					}
+					if (observationUnitRow.getEnvironmentVariables().containsKey(field.getName())) {
+						row.put(requiredField, observationUnitRow.getEnvironmentVariables().get(field.getName()).getValue());
+					}
+
+				} else {
+					if (STATIC_LOT_FIELD_IDS.contains(field.getId()) || STATIC_TRANSACTION_FIELD_IDS.contains(field.getId())) {
```

Comment:
I find ore legible if we do not have this OR here and just have an if section per list